### PR TITLE
Access profiles from posts and leaderboard

### PR DESCRIPTION
**Pull request by lucamouchel**
This PR deals with accessing users' profiles easily by clicking on the profile picture or username in the home feed or in the leaderboard to access the users' profile

In this PR, because the class DisplayUserProfileActivtiy is launched from different places, we do not have access to the Profile object from a post or the leaderboard, instead we have access to Uids. So I removed the object selectedProfile from the profile utils class and instead we should put the uid as an extra when opening a profile page.

### CODE REVIEW

File: app/src/main/java/ch/epfl/culturequest/ui/leaderboard/LeaderboardRecycleViewAdapter.java [hunk 1]
```diff
@@ -22,6 +27,7 @@
 public class LeaderboardRecycleViewAdapter extends RecyclerView.Adapter<LeaderboardRecycleViewAdapter.LeaderboardViewHolder> {
     private List<String> topNUserNames;
 
+    private List<String> topNUids;
     private List<String> topNUserScores;
```
Comment (hmiranda-queiros) [inline app/src/main/java/ch/epfl/culturequest/ui/leaderboard/LeaderboardRecycleViewAdapter.java:31]: Don't forget to add this field to the getItemCount() method logic

File: app/src/main/java/ch/epfl/culturequest/ui/profile/ProfileViewModel.java [hunk 2]
```diff
@@ -33,28 +33,31 @@ public ProfileViewModel(String uid) {
...
+            if (!profile.getUid().equals(uid)) {
+                Database.getProfile(uid).whenComplete((selectedProfile, e) -> {
+                    this.selectedProfile.setValue(selectedProfile);
+                    username.setValue(selectedProfile.getUsername());
+                    profilePictureUri.setValue(selectedProfile.getProfilePicture());
+                    // We load all the posts for a user in 1 query to the database. Initially, I queried only 4 posts at
+                    // a time, but it is computationally more efficient to do 1 big query:
+                    //https://stackoverflow.com/questions/3910317/is-it-better-to-return-one-big-query-or-a-few-smaller-ones#:~:text=It%20is%20significantly%20faster%20to,the%20server%20more%20each%20time.
+                    CompletableFuture<List<Post>> profilePosts = Database.getPosts(selectedProfile.getUid());
+                    profilePosts.handle((posts, t) -> {
```
Comment (hmiranda-queiros) [inline app/src/main/java/ch/epfl/culturequest/ui/profile/ProfileViewModel.java:49]: Is there a reason why you don't use "getValue()" here to get the Uid of the MutableData selectedProfile ?

**Review by hmiranda-queiros — CHANGES_REQUESTED** [review]
Good work ! I just added two comments on things you may have forgotten to add :)

File: app/src/main/java/ch/epfl/culturequest/ui/leaderboard/LeaderboardRecycleViewAdapter.java (hunk 1 above)
Comment (lucamouchel) [reply app/src/main/java/ch/epfl/culturequest/ui/leaderboard/LeaderboardRecycleViewAdapter.java:31]: Yes thanks for noticing I forgot to, please check the new implementation

File: app/src/main/java/ch/epfl/culturequest/ui/profile/ProfileViewModel.java (hunk 2 above)
Comment (lucamouchel) [reply app/src/main/java/ch/epfl/culturequest/ui/profile/ProfileViewModel.java:49]: yes, if you check the few lines over these ones, it calls the Database: ```Database.getProfile(uid).whenComplete((selectedProfile, e) -> { ...```
so the selected profile you mention is the one fetched from the database not the mutable live data. In that block i write `this.selectedProfile.setValue(selectedProfile);` so i could either use the mutable live data and use `getValue().getUid()` or just use the profile fetched from the database

File: app/src/main/java/ch/epfl/culturequest/ui/leaderboard/LeaderboardRecycleViewAdapter.java (hunk 1 above)
Comment (hmiranda-queiros) [reply app/src/main/java/ch/epfl/culturequest/ui/leaderboard/LeaderboardRecycleViewAdapter.java:31]: All the changes are good !

**Review by hmiranda-queiros — APPROVED** [review]
All good to merge !

**Pull request merged**